### PR TITLE
[AutoDiff] forbid derivative registration using @differentiable

### PR DESCRIPTION
**Pull request by marcrasi**
Forbids the `jvp:` and `vjp:` arguments to `@differentiable`, deletes all the logic that handles them, and updates tests to use `@derivative(of:)` instead.

I left an error message saying that `jvp:` and `vjp:` are deprecated, so that there is a good explanation for anyone who missed the deprecation warning and whose code stops working. I think we should wait until next release to remove this.

I have preemptively merged this into `tensorflow` because it causes quite a few problems there (more tests need to be updated). PR for that coming soon.

### CODE REVIEW

**Comment by marcrasi** [comment]
@swift-ci please test

**Review by rxwei — APPROVED** [review]
🎉 the moment has come!

File: utils/gyb_syntax_support/NodeSerializationCodes.py
Comment (dan-zheng) [inline utils/gyb_syntax_support/NodeSerializationCodes.py]: minor: I wonder if it's a breaking change to change syntax node serialization codes. If not, decrementing higher serialization codes would be nice, so there's no gap.

@rintaro: could you please provide a suggestion here?

**Review by dan-zheng — APPROVED** [review]
Yay!

---

When syntax definitions are changed, I believe `apple/swift-syntax` needs to be updated with newly generated syntax files. This is necessary to pass [`--swiftsyntax-verify-generated-files`](https://github.com/apple/swift/blob/7802d1755cea1e5e6207e12f92bf036d7c490b73/utils/build_swift/build_swift/driver_arguments.py#L570).

[Here's an example](https://github.com/apple/swift-syntax/pull/188). I think you can regenerate `swift-syntax` files with `utils/build-script --llbuild --swiftpm --swiftsyntax`.

File: utils/gyb_syntax_support/NodeSerializationCodes.py
Comment (marcrasi) [reply utils/gyb_syntax_support/NodeSerializationCodes.py]: I saw other places above that did this, so I imitated those.

Comment (rintaro) [reply utils/gyb_syntax_support/NodeSerializationCodes.py]: They are already in `swift-5.2-branch`. So please keep higher codes as is.

**Comment by marcrasi** [comment]
For the swift-syntax changes, do I need to simultaneously merge this PR with the swift-syntax update? Is there some way I can run CI on the two PRs together so that I can see if they work together? @rintaro?

If I do need to do it atomically, I'll split the swift-syntax changes out of this PR and do it separately so that I don't need to do the simultaneous merge thing with such a massive PR.

**Comment by dan-zheng** [comment]
> For the swift-syntax changes, do I need to simultaneously merge this PR with the swift-syntax update? Is there some way I can run CI on the two PRs together so that I can see if they work together? @rintaro?

Yes, you can run CI on the two PRs together.
Provide friend PR links _above_ the `@swift-ci` invocation. [Here's an example.](https://github.com/apple/swift/pull/28321#issuecomment-563904987)

**Comment by marcrasi** [comment]
Ok, I have rebased this and removed the syntax changes (will make them in a subsequent PR after merging this one). Running tests now.

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by marcrasi** [comment]
Rebased again. We're actually ready to merge this now. Running tests one more time.

**Comment by marcrasi** [comment]
@swift-ci please test

**Comment by dan-zheng** [comment]
Merging to unblock progress (on TF-836 `@differentiable` attribute serialization, which involves overlapping changes).

**Pull request merged**
